homeworks/06-redux-advanced/src/logSender.ts

```diff
@@ -0,0 +1,60 @@
+import { Dispatch, Middleware } from "redux";
+import { PizzaAction, PizzaActionType } from "./reducer";
+import { State } from "./types";
+
+const logSender: Middleware<{}, State, Dispatch<PizzaAction>> = store => next => action => {
+  const event = createLogEvent(action, store.getState())
+  console.log("Log event", event)
+  sendToServer(event)
+  next(action)
+}
+
+export default logSender
+
+type LogEvent = {
+  eventName: string,
+  pizzaName?: string,
+  pizzaPrice?: number
+}
+
+const eventNames = new Map([
+  [PizzaActionType.LOAD_LIST, "PIZZA_LIST_VIEWED"],
+  [PizzaActionType.ADD_TO_BASKET, "PIZZA_ADDED_TO_BASKET"],
+  [PizzaActionType.REMOVE_FROM_BASKET, "PIZZA_REMOVED_FROM_BASKET"]
+])
+const UnknownEventName = "UNKNOWN_EVENT"
+
+function createLogEvent(action: PizzaAction, state: State): LogEvent {
+  const eventName = eventNames.get(action.type) ?? UnknownEventName
+  const pizzaId = action.pizzaId
+  const pizza = pizzaId ? state.pizza.find(_ => _._id === pizzaId) : undefined
+  let event
+  if (pizza) {
```

Comment:
there can be just.
```
const pizza = state.find(p => p.id === pizzaId);
const event = {
  eventName: eventName,
};
if (pizza) {
  event.pizzaName = pizza.name;
  event.pizzaPrice = pizza.price;
}
```
`event` is object. so, we are able to add new fields.

Reply:
It doesn't work, show an error

> Property 'pizzaName' does not exist on type '{ eventName: string; }'.ts(2339)